File: sokol_examples/mipmap-sapp/mipmap-sapp.zig
//------------------------------------------------------------------------------
//  https://github.com/floooh/sokol-samples/blob/master/sapp/mipmap-sapp.c
//  Demonstrate all the mipmapping filters.
//------------------------------------------------------------------------------
const std = @import("std");
const sokol = @import("sokol");
const sg = sokol.gfx;
const dbgui = @import("dbgui");
const mipmap_shader = @import("mipmap-sapp.glsl.zig");
const rowmath = @import("rowmath");
const Mat4 = rowmath.Mat4;

const state = struct {
    var pip = sg.Pipeline{};
    var vbuf = sg.Buffer{};
    var tex_view = sg.View{};
    var smp: [12]sg.Sampler = undefined;
    var r: f32 = 0;
    const pixels = struct {
        var mip0: [65536]u32 = undefined; // 256x256
        var mip1: [16384]u32 = undefined; // 128x128
        var mip2: [4096]u32 = undefined; // 64*64
        var mip3: [1024]u32 = undefined; // 32*32
        var mip4: [256]u32 = undefined; // 16*16
        var mip5: [64]u32 = undefined; // 8*8
        var mip6: [16]u32 = undefined; // 4*4
        var mip7: [4]u32 = undefined; // 2*2
        var mip8: [1]u32 = undefined; // 1*2
    };
};

const mip_colors = [9]u32{
    0xFF0000FF, // red
    0xFF00FF00, // green
    0xFFFF0000, // blue
    0xFFFF00FF, // magenta
    0xFFFFFF00, // cyan
    0xFF00FFFF, // yellow
    0xFFFF00A0, // violet
    0xFFFFA0FF, // orange
    0xFFA000FF, // purple
};

export fn init() void {
    sg.setup(.{
        .environment = sokol.glue.environment(),
        .logger = .{ .func = sokol.log.func },
    });
    dbgui.setup(sokol.app.sampleCount());

    // a plane vertex buffer
    const vertices = [_]f32{
        -1.0, -1.0, 0.0, 0.0, 0.0,
        1.0,  -1.0, 0.0, 1.0, 0.0,
        -1.0, 1.0,  0.0, 0.0, 1.0,
        1.0,  1.0,  0.0, 1.0, 1.0,
    };
    state.vbuf = sg.makeBuffer(.{
        .data = sg.asRange(&vertices),
    });

    // create image with mipmap content, different colors and checkboard pattern
    var img_data = sg.ImageData{};
    var even_odd = false;
    inline for (0..9) |mip_index| {
        var ptr: [*]u32 = switch (mip_index) {
            0 => @ptrCast(&state.pixels.mip0[0]),
            1 => @ptrCast(&state.pixels.mip1[0]),
            2 => @ptrCast(&state.pixels.mip2[0]),
            3 => @ptrCast(&state.pixels.mip3[0]),
            4 => @ptrCast(&state.pixels.mip4[0]),
            5 => @ptrCast(&state.pixels.mip5[0]),
            6 => @ptrCast(&state.pixels.mip6[0]),
            7 => @ptrCast(&state.pixels.mip7[0]),
            8 => @ptrCast(&state.pixels.mip8[0]),
            else => unreachable,
        };
        const dim = 1 << (8 - mip_index);
        img_data.mip_levels[mip_index].ptr = ptr;
        img_data.mip_levels[mip_index].size = dim * dim * 4;
        for (0..dim) |_| {
            for (0..dim) |_| {
                ptr[0] = if (even_odd) mip_colors[mip_index] else 0xFF000000;
                ptr = @ptrCast(&ptr[1]);
                even_odd = !even_odd;
            }
            even_odd = !even_odd;
        }
    }
    const img = sg.makeImage(.{
        .width = 256,
        .height = 256,
        .num_mipmaps = 9,
        .pixel_format = .RGBA8,
        .data = img_data,
    });
    state.tex_view = sg.makeView(.{
        .texture = .{
            .image = img,
        },
    });

    // the first 4 samplers are just different min-filters
    var smp_desc = sg.SamplerDesc{
        .mag_filter = .LINEAR,
    };
    const filters = [2]sg.Filter{
        .NEAREST,
        .LINEAR,
    };
    const mipmap_filters = [2]sg.Filter{
        .NEAREST,
        .LINEAR,
    };
    var smp_index: usize = 0;
    for (0..2) |i| {
        for (0..2) |j| {
            smp_desc.min_filter = filters[i];
            smp_desc.mipmap_filter = mipmap_filters[j];
            state.smp[smp_index] = sg.makeSampler(smp_desc);
            smp_index += 1;
        }
    }
    // the next 4 samplers use min_lod/max_lod
    smp_desc.min_lod = 2.0;
    smp_desc.max_lod = 4.0;
    for (0..2) |i| {
        for (0..2) |j| {
            smp_desc.min_filter = filters[i];
            smp_desc.mipmap_filter = mipmap_filters[j];
            state.smp[smp_index] = sg.makeSampler(smp_desc);
            smp_index += 1;
        }
    }
    // the last 4 samplers use different anistropy levels
    smp_desc.min_lod = 0.0;
    smp_desc.max_lod = 0.0; // for max_lod, zero-initialized means "FLT_MAX"
    smp_desc.min_filter = .LINEAR;
    smp_desc.mag_filter = .LINEAR;
    smp_desc.mipmap_filter = .LINEAR;
    inline for (0..4) |i| {
        smp_desc.max_anisotropy = 1 << i;
        state.smp[smp_index] = sg.makeSampler(smp_desc);
        smp_index += 1;
    }
    std.debug.assert(smp_index == 12);

    // pipeline state
    var pip_desc = sg.PipelineDesc{
        .shader = sg.makeShader(
            mipmap_shader.mipmapShaderDesc(sg.queryBackend()),
        ),
        .primitive_type = .TRIANGLE_STRIP,
    };
    pip_desc.layout.attrs[mipmap_shader.ATTR_mipmap_pos].format = .FLOAT3;
    pip_desc.layout.attrs[mipmap_shader.ATTR_mipmap_uv0].format = .FLOAT2;
    state.pip = sg.makePipeline(pip_desc);
}

export fn frame() void {
    defer sg.commit();

    const proj = Mat4.makePerspective(
        std.math.degreesToRadians(90.0),
        sokol.app.widthf() / sokol.app.heightf(),
        0.01,
        10.0,
    );
    const view = Mat4.makeLookAt(
        .{ .x = 0.0, .y = 0.0, .z = 5.0 },
        .{ .x = 0.0, .y = 0.0, .z = 0.0 },
        .{ .x = 0.0, .y = 1.0, .z = 0.0 },
    );
    const view_proj = view.mul(proj);

    state.r += @floatCast(0.1 * 60.0 * sokol.app.frameDuration());
    const rm = Mat4.makeRotation(state.r, .{ .x = 1.0, .y = 0.0, .z = 0.0 });

    var bind = sg.Bindings{};
    bind.vertex_buffers[0] = state.vbuf;
    bind.views[mipmap_shader.VIEW_tex] = state.tex_view;
    {
        sg.beginPass(.{ .swapchain = sokol.glue.swapchain() });
        defer sg.endPass();
        sg.applyPipeline(state.pip);
        for (0..12) |i| {
            const x = (@as(f32, @floatFromInt(i & 3)) - 1.5) * 2.0;
            const y = (@as(f32, @floatFromInt(i / 4)) - 1.0) * -2.0;
            const model = rm.mul(Mat4.makeTranslation(.{ .x = x, .y = y, .z = 0.0 }));
            const vs_params = mipmap_shader.VsParams{
                .mvp = model.mul(view_proj).m,
            };

            bind.samplers[mipmap_shader.SMP_smp] = state.smp[i];
            sg.applyBindings(bind);
            sg.applyUniforms(
                mipmap_shader.UB_vs_params,
                sg.asRange(&vs_params),
            );
            sg.draw(0, 4, 1);
        }
        dbgui.draw();
    }
}

export fn cleanup() void {
    dbgui.shutdown();
    sg.shutdown();
}

pub fn main() void {
    sokol.app.run(.{
        .init_cb = init,
        .frame_cb = frame,
        .cleanup_cb = cleanup,
        .event_cb = dbgui.event,
        .width = 800,
        .height = 600,
        .sample_count = 4,
        .window_title = "mipmap-sapp.c",
        .icon = .{ .sokol_default = true },
        .logger = .{ .func = sokol.log.func },
    });
}

File: sokol_examples/mipmap-sapp/mipmap-sapp.c
//------------------------------------------------------------------------------
//  mipmap-sapp.c
//  Demonstrate all the mipmapping filters.
//------------------------------------------------------------------------------
#include "sokol_app.h"
#include "sokol_gfx.h"
#include "sokol_log.h"
#include "sokol_glue.h"
#define HANDMADE_MATH_IMPLEMENTATION
#define HANDMADE_MATH_NO_SSE
#include "HandmadeMath.h"
#include "dbgui/dbgui.h"
#include "mipmap-sapp.glsl.h"
#include <assert.h>

static struct {
    sg_pipeline pip;
    sg_buffer vbuf;
    sg_image img;
    sg_sampler smp[12];
    float r;
    struct {
        uint32_t mip0[65536];   // 256x256
        uint32_t mip1[16384];   // 128x128
        uint32_t mip2[4096];    // 64*64
        uint32_t mip3[1024];    // 32*32
        uint32_t mip4[256];     // 16*16
        uint32_t mip5[64];      // 8*8
        uint32_t mip6[16];      // 4*4
        uint32_t mip7[4];       // 2*2
        uint32_t mip8[1];       // 1*2
    } pixels;
} state;

static const uint32_t mip_colors[9] = {
    0xFF0000FF,     // red
    0xFF00FF00,     // green
    0xFFFF0000,     // blue
    0xFFFF00FF,     // magenta
    0xFFFFFF00,     // cyan
    0xFF00FFFF,     // yellow
    0xFFFF00A0,     // violet
    0xFFFFA0FF,     // orange
    0xFFA000FF,     // purple
};

void init(void) {
    sg_setup(&(sg_desc){
        .environment = sglue_environment(),
        .logger.func = slog_func,
    });
    __dbgui_setup(sapp_sample_count());

    // a plane vertex buffer
    float vertices[] = {
        -1.0, -1.0, 0.0,  0.0, 0.0,
        +1.0, -1.0, 0.0,  1.0, 0.0,
        -1.0, +1.0, 0.0,  0.0, 1.0,
        +1.0, +1.0, 0.0,  1.0, 1.0,
    };
    state.vbuf = sg_make_buffer(&(sg_buffer_desc){
        .data = SG_RANGE(vertices)
    });

    // create image with mipmap content, different colors and checkboard pattern
    sg_image_data img_data;
    uint32_t* ptr = state.pixels.mip0;
    bool even_odd = false;
    for (int mip_index = 0; mip_index <= 8; mip_index++) {
        const int dim = 1<<(8-mip_index);
        img_data.subimage[0][mip_index].ptr = ptr;
        img_data.subimage[0][mip_index].size = (size_t) (dim * dim * 4);
        for (int y = 0; y < dim; y++) {
            for (int x = 0; x < dim; x++) {
                *ptr++ = even_odd ? mip_colors[mip_index] : 0xFF000000;
                even_odd = !even_odd;
            }
            even_odd = !even_odd;
        }
    }
    state.img = sg_make_image(&(sg_image_desc){
        .width = 256,
        .height = 256,
        .num_mipmaps = 9,
        .pixel_format = SG_PIXELFORMAT_RGBA8,
        .data = img_data
    });

    // the first 4 samplers are just different min-filters
    sg_sampler_desc smp_desc = {
        .mag_filter = SG_FILTER_LINEAR,
    };
    sg_filter filters[] = {
        SG_FILTER_NEAREST,
        SG_FILTER_LINEAR,
    };
    sg_filter mipmap_filters[] = {
        SG_FILTER_NEAREST,
        SG_FILTER_LINEAR,
    };
    int smp_index = 0;
    for (int i = 0; i < 2; i++) {
        for (int j = 0; j < 2; j++) {
            smp_desc.min_filter = filters[i];
            smp_desc.mipmap_filter = mipmap_filters[j];
            state.smp[smp_index++] = sg_make_sampler(&smp_desc);
        }
    }
    // the next 4 samplers use min_lod/max_lod
    smp_desc.min_lod = 2.0f;
    smp_desc.max_lod = 4.0f;
    for (int i = 0; i < 2; i++) {
        for (int j = 0; j < 2; j++) {
            smp_desc.min_filter = filters[i];
            smp_desc.mipmap_filter = mipmap_filters[j];
            state.smp[smp_index++] = sg_make_sampler(&smp_desc);
        }
    }
    // the last 4 samplers use different anistropy levels
    smp_desc.min_lod = 0.0f;
    smp_desc.max_lod = 0.0f;    // for max_lod, zero-initialized means "FLT_MAX"
    smp_desc.min_filter = SG_FILTER_LINEAR;
    smp_desc.mag_filter = SG_FILTER_LINEAR;
    smp_desc.mipmap_filter = SG_FILTER_LINEAR;
    for (int i = 0; i < 4; i++) {
        smp_desc.max_anisotropy = 1<<i;
        state.smp[smp_index++] = sg_make_sampler(&smp_desc);
    }
    assert(smp_index == 12);

    // pipeline state
    state.pip = sg_make_pipeline(&(sg_pipeline_desc) {
        .layout = {
            .attrs = {
                [ATTR_vs_pos].format = SG_VERTEXFORMAT_FLOAT3,
                [ATTR_vs_uv0].format = SG_VERTEXFORMAT_FLOAT2
            }
        },
        .shader = sg_make_shader(mipmap_shader_desc(sg_query_backend())),
        .primitive_type = SG_PRIMITIVETYPE_TRIANGLE_STRIP,
    });
}

void frame(void) {
    hmm_mat4 proj = HMM_Perspective(90.0f, sapp_widthf()/sapp_heightf(), 0.01f, 10.0f);
    hmm_mat4 view = HMM_LookAt(HMM_Vec3(0.0f, 0.0f, 5.0f), HMM_Vec3(0.0f, 0.0f, 0.0f), HMM_Vec3(0.0f, 1.0f, 0.0f));
    hmm_mat4 view_proj = HMM_MultiplyMat4(proj, view);
    vs_params_t vs_params;
    state.r += 0.1f * 60.0f * (float)sapp_frame_duration();
    hmm_mat4 rm = HMM_Rotate(state.r, HMM_Vec3(1.0f, 0.0f, 0.0f));

    sg_bindings bind = {
        .vertex_buffers[0] = state.vbuf,
        .fs.images[SLOT_tex] = state.img,
    };
    sg_begin_pass(&(sg_pass){ .swapchain = sglue_swapchain() });
    sg_apply_pipeline(state.pip);
    for (int i = 0; i < 12; i++) {
        const float x = ((float)(i & 3) - 1.5f) * 2.0f;
        const float y = ((float)(i / 4) - 1.0f) * -2.0f;
        hmm_mat4 model = HMM_MultiplyMat4(HMM_Translate(HMM_Vec3(x, y, 0.0f)), rm);
        vs_params.mvp = HMM_MultiplyMat4(view_proj, model);

        bind.fs.samplers[SLOT_smp] = state.smp[i];
        sg_apply_bindings(&bind);
        sg_apply_uniforms(SG_SHADERSTAGE_VS, SLOT_vs_params, &SG_RANGE(vs_params));
        sg_draw(0, 4, 1);
    }
    __dbgui_draw();
    sg_end_pass();
    sg_commit();
}

void cleanup(void) {
    __dbgui_shutdown();
    sg_shutdown();
}

sapp_desc sokol_main(int argc, char* argv[]) {
    (void)argc; (void)argv;
    return (sapp_desc){
        .init_cb = init,
        .frame_cb = frame,
        .cleanup_cb = cleanup,
        .event_cb = __dbgui_event,
        .width = 800,
        .height = 600,
        .sample_count = 4,
        .window_title = "mipmap-sapp.c",
        .icon.sokol_default = true,
        .logger.func = slog_func,
    };
}
